### PR TITLE
Segment perf length

### PR DESCRIPTION
**Pull request by joakim-hove**
Refactorings to get rid of unwanted manipulation of ICD object in opm-simulator.

The main content of this PR is the new COMPSEGS handling in commit: "Refactor COMPSEGS handling"

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2684 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2648 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2648 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2648 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2648 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2648 please

File: src/opm/parser/eclipse/EclipseState/Schedule/MSW/Segment.cpp [hunk 1]
```diff
@@ -209,6 +209,9 @@ namespace {
         return m_depth;
     }
 
+    double Segment::perfLength() const {
+        return *this->m_perf_length;
```
Comment (GitPaean) [inline src/opm/parser/eclipse/EclipseState/Schedule/MSW/Segment.cpp:213]: is this `*` intentional?

**Comment by GitPaean** [comment]
I understand the major motivation of this PR. The big question to ask here is whether the processing of the grid is done when `handleWSEGSICD`.  My knowledge might be out-dated about this.

**Comment by joakim-hove** [comment]
> The big question to ask here is whether the processing of the grid (like PINCH) is done when handleWSEGSICD

Ehhh - well I understand the principal point in your statement but I don't agree:

1. This PR does not change the behavior of the `perf_length` with respect to PINCH operations - they were ignored before and are still ignored in this respect.
2. This must be an extremely small effect anyway - not something I would go after.

File: src/opm/parser/eclipse/EclipseState/Schedule/MSW/Segment.cpp (hunk 1 above)
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/MSW/Segment.cpp:213]: Yes - that is intentional; the `m_perf_length` member is a `std::optional<double>` as a way to safeguard against accessing an uninitialized value. A bit more verbose could be:
```C++
if (this->m_perf_length)
     return *this->m_perf_length;
throw std::logic_error("Tryng to access an uninitialized perforation length variable");
```

**Comment by GitPaean** [comment]
> This PR does not change the behavior of the perf_length with respect to PINCH operations - they were ignored before and are still ignored in this respect.

You are probably right, while I could not follow this here.  If a perforation is pinched out, will it affect the perf_length related to a segment?

**Comment by alfbr** [comment]
Do any of our test cases cover this?

**Comment by GitPaean** [comment]
> Do any of our test cases cover this?

We have some model 2 realizations that have perforated cell gets pinched out, that is why we have the function `filterConnections` in the `Schedule`, which is called in `ebos`. But there is no SICD involved there. 

We do not have a test case covers the corner situation, as far as I understand.

**Comment by joakim-hove** [comment]
> You are probably right, while I could not follow this here. If a perforation is pinched out, will it affect the perf_length related to a segment?

The treatment of perforation range is 100% identical in this PR and in master; furthermore this is akin to bikeshedding - the perforation length we are talking about here is taken directly from items 5 and 6 in the COMPSEGS keyword - this is actually grid independent input.

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2648 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2648 please

**Comment by GitPaean** [comment]
> The treatment of perforation range is 100% identical in this PR and in master; furthermore this is akin to bikeshedding - the perforation length we are talking about here is taken directly from items 5 and 6 in the COMPSEGS keyword - this is actually grid independent input.

I totally understand what you said. I just do not know what happens if a perforation associated to a certain segment is pinched out. That is why the implementation is called through the opm-simulators which is after the grid is processed. 

If you are correct, they it should be called in the opm-common in the first place.

**Comment by GitPaean** [comment]
I am not against the proposal here. Just commenting out for discussion. 

If we found out differently later, we can always fix it.

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2648 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2648 please

**Comment by joakim-hove** [comment]
OK - the refactorings have been merged; this along with opm-simulators/2648 is the real change. Will merge later today if Green.

**Pull request merged**